lisdf/parsing/xml_j/visitor.py

```diff
@@ -0,0 +1,138 @@
+#! /usr/bin/env python3
+# -*- coding: utf-8 -*-
+# File   : visitor.py
+# Author : Jiayuan Mao
+# Email  : maojiayuan@gmail.com
+# Date   : 03/23/2022
+#
+# This file is part of lisdf.
+# Distributed under terms of the MIT license.
+
+import functools
+import os.path as osp
+from collections import defaultdict
+from typing import Any, Callable, Dict, List, Optional
+
+from lisdf.utils.printing import indent_text
+
+from .xml import XMLNode, load_file
+
+__all__ = ["XMLVisitor"]
+
+
+class XMLVisitor(object):
+    def __init__(self) -> None:
+        self.filename_stack: List[str] = list()
+        self.node_stack: List[XMLNode] = list()
+        self._st: Dict[str, List[Any]] = defaultdict(list)  # optional stacks.
+        self._data: Dict[str, Dict[str, Any]] = defaultdict(dict)  # optional data.
+        self._indent: int = 0
+
+        self.verbose = False
+
+    def set_verbose(self, flag: bool = True) -> None:
+        self.verbose = flag
+
+    def _get_processor(self, tag: str) -> Optional[Callable[[XMLNode], Any]]:
+        tag = tag.replace("-", "_")
+        return getattr(self, tag, None)
+
+    def load_file(self, filename: str) -> Any:
+        node = load_file(filename)
+        return self.visit(filename, node)
+
+    def visit(self, filename: str, root: XMLNode) -> Any:
+        self.filename_stack.append(filename)
+        if self.verbose:
+            print(indent_text(filename, self._indent))
+
+        def _inner(node: XMLNode):
+            if self.verbose:
+                print(indent_text(node.open_tag(), self._indent))
+                if node.text:
+                    print(indent_text(node.text, self._indent + 1))
+            self._indent += 1
+            try:
+                proc = self._get_processor(node.tag + "_init")
+                if proc is not None:
+                    proc(node)
+                self.node_stack.append(node)
+                for i, c in enumerate(node.children):
+                    node.children[i] = _inner(c)
+                node.children = [c for c in node.children if c is not None]
+                self.node_stack.pop()
+                proc = self._get_processor(node.tag)
+
+                if proc is not None:
+                    if self.verbose:
+                        print(indent_text(node.close_tag(), self._indent - 1))
+                    return proc(node)
+                if self.verbose:
+                    print(indent_text(node.close_tag(), self._indent - 1))
+                return node
+            except Exception as e:
+                raise e
+            finally:
+                self._indent -= 1
+
+        try:
+            root = root.clone()
+            return _inner(root)
+        finally:
+            self.filename_stack.pop()
```

Comment:
same comment as above here - why does this code need to be wrappen in a `try/finally` block? is it really necessary?

Reply:
I admit that this is my way of writing `defer self.filename_stack.pop()` (otherwise I need to store the `rv = inner(root)`...

Reply:
done